src/network_service/sending_targets_cache.rs

```diff
@@ -52,48 +53,78 @@ impl SendingTargetsCache {
         let _ = self.cache.insert(msg_id, targets);
     }
 
+    fn target_states<'a>(
+        &'a self,
+        msg_id: u64,
+    ) -> impl Iterator<Item = &'a (ConnectionInfo, TargetState)> {
+        self.cache.get(&msg_id).into_iter().flatten()
+    }
+
+    fn target_states_mut<'a>(
+        &'a mut self,
+        msg_id: u64,
+    ) -> impl Iterator<Item = &'a mut (ConnectionInfo, TargetState)> {
+        self.cache.get_mut(&msg_id).into_iter().flatten()
+    }
+
     fn fail_target(&mut self, msg_id: u64, target: SocketAddr) {
-        for (tgt, state) in self.cache.get_mut(&msg_id).into_iter().flatten() {
-            if tgt.peer_addr() == target {
-                match *state {
-                    TargetState::Sending(x) | TargetState::Failed(x) => {
-                        *state = TargetState::Failed(x + 1);
-                    }
-                    TargetState::Sent => {
-                        panic!("A target that should no longer fail - failed!");
-                    }
+        let _ = self
+            .target_states_mut(msg_id)
+            .find(|(info, _state)| info.peer_addr() == target)
+            .map(|(_info, state)| match *state {
+                TargetState::Failed(_) => {
+                    error!("Got a failure from a failed target!");
                 }
-                break;
-            }
-        }
+                TargetState::Sending(x) => {
+                    *state = TargetState::Failed(x + 1);
+                }
+                TargetState::Sent => {
+                    log_or_panic!(
+                        LogLevel::Error,
+                        "A target that should no longer fail - failed!"
+                    );
+                }
+            });
     }
 
     fn get_next_target(&self, msg_id: u64) -> Option<ConnectionInfo> {
-        let mut best_target = None;
-        let mut min_num = MAX_RESENDS;
-        for (tgt, status) in self.cache.get(&msg_id).into_iter().flatten() {
-            match status {
-                TargetState::Failed(x) if *x < min_num => {
-                    best_target = Some(tgt);
-                    min_num = *x;
-                }
-                _ => (),
-            }
-        }
-        best_target.cloned()
+        self.target_states(msg_id)
+            .filter_map(|(info, state)| match state {
+                TargetState::Failed(x) if *x <= MAX_RESENDS => Some((info, *x)),
+                _ => None,
+            })
+            .min_by_key(|&(_info, num)| num)
+            .map(|(info, _num)| info)
+            .cloned()
+    }
+
+    fn should_drop(&self, msg_id: u64) -> bool {
+        self.target_states(msg_id)
+            .all(|(_info, state)| match state {
+                TargetState::Sent => true,
+                TargetState::Failed(x) if *x > MAX_RESENDS => true,
```

Comment:
Not 100% happy about that condition being express twice, in different ways as well.
`if *x > MAX_RESENDS` `if *x <= MAX_RESENDS`.

Wondering maybe something like:

```rust
     fn is_complete(&self) -> bool {
          match state {
                TargetState::Sent => true,
                TargetState::Failed(x) => *x > MAX_RESENDS,
                TargetState::Sending => false,
         }
     }

     fn get_next_target(&self, msg_id: u64) -> Option<ConnectionInfo> {
        self.target_states(msg_id)
            .filter(|(info, state)| !state.is_complete())
            .filter_map(|(info, state)| state {
                TargetState::Failed(x) => Some((info, *x)),
                _ => None,
            })
            .min_by_key(|&(_info, num)| num)
            .map(|(info, _num)| info)
            .cloned()
    }

     fn should_drop(&self, msg_id: u64) -> bool {
        self.target_states(msg_id).all(|(_info, state)| state.is_complete())
    } 
```